src/shared/auth/auth.controller.spec.ts

```diff
@@ -1,22 +1,390 @@
+      return Promise.resolve({
+        emailVerificationCode: '123456',
+        emailVerificationUrl: 'http://localhost/verify/123456',
+      });
+    }),
+    findByUsername: jest.fn(),
+    validateUser: jest.fn(),
+  };
+  // mock for the static methods in the user schema
+  const mockUserModel = {
+    verifyEmail: jest
+      .fn()
+      .mockImplementation((userId: string, token: string) => {
+        if (token === '123456') {
+          return Promise.resolve({
+            status: 200,
+            message: 'Email Verification Successful',
+          });
+        }
+        return Promise.reject(
+          new BadRequestException('Invalid email verification token supplied'),
+        );
+      }),
+
+    signUp: jest
+      .fn()
+      .mockImplementation((req: any, createUserDto: any, sso = false) => {
+        return Promise.resolve({
+          ...mockUser,
+          ...createUserDto,
+          emailVerificationCode: '123456',
+          emailVerificationUrl: 'http://localhost/verify/123456',
+        });
+      }),
+
+    forgetPassword: jest.fn().mockImplementation((email: string) => {
+      if (email === 'john@example.com') {
+        return Promise.resolve({
+          email: 'john@example.com',
+          firstName: 'John',
+          lastName: 'Doe',
+        });
+      }
+      return Promise.reject(
+        new BadRequestException('Email supplied cannot be found'),
+      );
+    }),
+
+    sendEmailVerificationToken: jest.fn().mockImplementation((req, userId) => {
+      return Promise.resolve({
+        emailVerificationCode: '123456',
+        emailVerificationUrl: 'http://localhost/verify/123456',
+      });
+    }),
+
+    // Base mongoose model methods
+    generateUserHandle: jest.fn(),
+    findOne: jest.fn(),
+    findById: jest.fn(),
+    create: jest.fn(),
+  } as unknown as UserModel; // bypass typescript type checking || i know... but it works
+
+  // ensure that the jwt always returns the same token
+  const mockJwtService = {
+    sign: jest.fn().mockReturnValue('mock_token'),
+  };
 
   beforeEach(async () => {
+    jest.clearAllMocks(); // clear all mock implementaions
+
     const module: TestingModule = await Test.createTestingModule({
-      imports: [TestModule],
       controllers: [AuthController],
-      providers: [AuthService],
+      providers: [
+        {
+          provide: AuthService,
+          useValue: mockAuthService,
+        },
+        {
+          provide: User.name,
+          useValue: mockUserModel,
+        },
+        {
+          provide: JwtService,
+          useValue: mockJwtService,
+        },
+      ],
     }).compile();
 
     controller = module.get<AuthController>(AuthController);
+    userModel = module.get<UserModel>(User.name);
+    authService = module.get<MockAuthService>(AuthService);
+    jwtService = module.get<JwtService>(JwtService);
   });
 
+  // check if controller is created
   it('should be defined', () => {
     expect(controller).toBeDefined();
```

Comment:
What controller is this, you can use the right name.

Reply:
Renamed for better representation in new commit